File: src/main/java/io/github/shirohoo/library/application/service/user/UserService.java
package io.github.shirohoo.library.application.service.user;

import io.github.shirohoo.library.domain.user.User;
import io.github.shirohoo.library.domain.user.UserRepository;
import java.util.List;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
public class UserService {
    private final UserRepository userRepository;

    public UserService(UserRepository userRepository) {
        this.userRepository = userRepository;
    }

    @Transactional
    public void saveUser(User newUser) {
        userRepository.save(newUser);
    }

    @Transactional(readOnly = true)
    public List<User> getUsers() {
        return userRepository.findAll();
    }

    @Transactional
    public void updateUsername(long userId, String username) {
        User user = userRepository.findBy(userId).orElseThrow();
        user.updateName(username);
    }

    @Transactional
    public void deleteUser(String name) {
        User user = userRepository.findBy(name).orElseThrow();
        userRepository.delete(user);
    }
}

File: src/main/kotlin/io/github/shirohoo/library/application/service/user/UserService.kt
package io.github.shirohoo.library.application.service.user

import io.github.shirohoo.library.domain.user.User
import io.github.shirohoo.library.domain.user.UserRepository
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class UserService(
    private val userRepository: UserRepository
) {
    @Transactional
    fun saveUser(newUser: User) = userRepository.save(newUser)

    @Transactional(readOnly = true)
    fun getUsers(): List<User> = userRepository.findAll()

    @Transactional
    fun updateUsername(userId: Long, username: String) {
        val user = userRepository.findBy(userId)
        user.updateName(username)
    }

    @Transactional
    fun deleteUser(username: String) {
        val user = userRepository.findBy(username)
        user.apply(userRepository::delete)
    }

    fun findAllWithHistories(): List<User> = userRepository.findAllWithHistories()
}